publish C# 7.0 proposals

BillWagner opened the pull request:
Fixes #10721

Internal review links:

[Pattern matching](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-7.0/pattern-matching?branch=pr-en-us-11026)
[Local functions](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-7.0/local-functions?branch=pr-en-us-11026)
[out variable declarations](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-7.0/out-var?branch=pr-en-us-11026)
[throw expressions](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-7.0/throw-expression?branch=pr-en-us-11026)
[binary literals](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-7.0/binary-literals?branch=pr-en-us-11026)
[digit separators](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-7.0/digit-separators?branch=pr-en-us-11026)

/cc @MadsTorgersen

BillWagner committed:
publish C# 7.0 proposals

Fixes #10721

BillWagner committed:
fix build issues

BillWagner committed:
update incorrect link.

BillWagner committed:
refine TOC headers

BillWagner commented:
dotnet/csharplang#2296 will fix the one build warnings introduced in this PR.

BillWagner committed:
add breadcrumbs

mairaw commented on .openpublishing.redirection.json:
```diff
@@ -2394,6 +2394,10 @@
+        },
+        {
+            "source_path": "docs/csharp/language-reference/language-specification/index.md",
+            "redirect_url": "docs/csharp/language-reference/language-specification/introduction/",
```
```suggestion
            "redirect_url": "/dotnet/csharp/language-reference/language-specification/introduction/",
```

mairaw commented on docfx.json:
```diff
@@ -155,6 +169,7 @@
       },
       "ms.author": {
         "_csharplang/spec/*.md": "wiwagn",
+        "_csharplang/proposals/csharp-7.0/*.md": "wiwagn",
```
you could perhaps just change the previous line to 
"_csharplang/**/**.md": "wiwagn",

BillWagner replied:
@mairaw I want to wait on that until the next updates (7.1 or 7.2). I think that's where it's going, but I need to understand what other content from the dependent repo could be accessed through docs. (It's all public, so it's not a danger, but it could mislead customers to include "rejected" or other proposals that aren't delivered yet.)

pkulikov commented on docs/csharp/language-reference/toc.md:
```diff
@@ -7,4 +7,4 @@
 ## [C# Compiler Options](compiler-options/)
 ## [C# Compiler Errors](compiler-messages/)
-## [C# 6.0 draft Language Specification](language-specification/)
+## [C# Language Specification and proposals](language-specification/)
```
Should the whole title be sentence-cased?

pkulikov commented on docs/csharp/language-reference/language-specification/index.md:
```diff
@@ -1,29 +0,0 @@
-
-The C# language specification is the definitive source for C# syntax and usage. This specification contains detailed information about all aspects of the language, including many points that the documentation for C# doesn't cover.
-
-Version 5.0 of the specification has been released in December 2017 as the [Standard ECMA-334 5th Edition](https://www.ecma-international.org/publications/files/ECMA-ST/ECMA-334.pdf) document.
```
With this removal, the link to the official latest spec is gone. Or that's fine, because the C# 6 draft includes the related content?

BillWagner replied:
I agree. I moved this text into the language reference index page.

mairaw committed:
Update .openpublishing.redirection.json

Co-Authored-By: BillWagner <wiwagn@microsoft.com>

BillWagner committed:
respond to feedback

BillWagner committed:
Attempt to grab all proposals that should be published.

BillWagner commented:
The build warnings should be addressed by dotnet/csharplang#2301

pkulikov commented on .openpublishing.redirection.json:
```diff
@@ -2404,6 +2404,10 @@
             "redirect_document_id": true
+        },
+        {
+            "source_path": "docs/csharp/language-reference/language-specification/index.md",
```
nit-pick: the entry is not in alphabetical order by the `source_path`
recently @mairaw has ordered this file, but I see that entry above is also not in "its" spot.

BillWagner replied:
I also moved the workflow entry to the correct location.

pkulikov commented on docfx.json:
```diff
@@ -33,6 +33,41 @@
+        "exclude": [
+          "README.md",
+          "LICENSE.md",
+          "proposal-template.md",
```
"proposal-template.md" entry appears twice in this list (line 59)

BillWagner committed:
respond to feedback

BillWagner commented:
closing and reopening to start a new build.

BillWagner commented:
This is ready for a final review.   

/cc @mairaw

mairaw commented on docs/csharp/language-reference/language-specification/toc.md:
```diff
@@ -18,3 +18,10 @@
+## [out variable declarations](../../../../_csharplang/proposals/csharp-7.0/out-var.md)
+## [throw expressions](../../../../_csharplang/proposals/csharp-7.0/throw-expression.md)
+## [binary literals](../../../../_csharplang/proposals/csharp-7.0/binary-literals.md)
+## [digit separators](../../../../_csharplang/proposals/csharp-7.0/digit-separators.md)
```
Should these entries use sentence case?

```suggestion
## [Digit separators](../../../../_csharplang/proposals/csharp-7.0/digit-separators.md)
```

mairaw commented on docfx.json:
```diff
@@ -33,6 +33,40 @@
+        "files": [
+          "**/*.md"
+        ],
+        "src": "_csharplang/proposals",
```
Should this be?

```suggestion
        "src": "_csharplang/proposals/csharp-7.0",
```

BillWagner replied:
I took at different path on this one.

mairaw commented on docs/csharp/language-reference/language-specification/toc.md:
```diff
@@ -18,3 +18,10 @@
 ## [Attributes](../../../../_csharplang/spec/attributes.md)
 ## [Unsafe code](../../../../_csharplang/spec/unsafe-code.md)
 ## [Documentation comments](../../../../_csharplang/spec/documentation-comments.md)
+# C# 7.0 language proposals 
```
shouldn't this be on its own TOC file under the proposals folder? Also a couple of paths are not working right now such as https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-7.0/?branch=pr-en-us-11026 or https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/?branch=pr-en-us-11026

It's fine if you don't want to do that, but then I think I'd add an entry above the draft and language proposals called "C# language specification and proposals" to match what they clicked on the TOC.

BillWagner replied:
I addressed this with redirects to the first result in each folder.

mairaw reviewed (approved):
LGTM but left a few comments for you to consider. The experience right now is a bit broken with this new addition and is also including some unwanted articles I believe. Feel free to ping me if you'd like to review this again after the changes are done.

mairaw committed:
Update docs/csharp/language-reference/language-specification/toc.md

Co-Authored-By: BillWagner <wiwagn@microsoft.com>

BillWagner committed:
clean up includes

1. redirect /proposals/csharp-7.0 and /proposals to first proposal doc.
1. update index to include link to proposals.
1. sentance case TOC for proposals.
1. limit which docs are published from the proposals directory

Testing if the ! syntax works to exclude an exclusion.

BillWagner committed:
next try on including only the right files.

BillWagner committed:
remove brackets on proposal

BillWagner committed:
provide a landing page for the spec and all proposals

BillWagner committed:
rename spec directory

BillWagner committed:
fix the spec and proposal TOC

BillWagner committed:
update breadcrumbs for new landing page.

mairaw commented:
Non-blocking suggestion: should you add a titlesuffix to those so they rank better for SEO?

mairaw commented on docfx.json:
```diff
@@ -29,10 +29,22 @@
+        "files": [
+          "records.md",
+          "csharp-7.0/*.md",
+          "csharp-8.0/patterns.md"
```
This one is not on the TOC: https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/patterns?branch=pr-en-us-11026

mairaw commented on docfx.json:
```diff
@@ -29,10 +29,22 @@
       },
+      {
+        "files": [
+          "records.md",
```
Same for this one as well. Not on TOC: https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/records?branch=pr-en-us-11026

BillWagner committed:
simplify directory structure

BillWagner committed:
add title suffix.

BillWagner committed:
fix redirects
